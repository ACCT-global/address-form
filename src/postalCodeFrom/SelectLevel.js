import React, { Component } from 'react'
import PropTypes from 'prop-types'
import AddressShapeWithValidation
  from '../propTypes/AddressShapeWithValidation'
import InputSelect from '../addressInputs/InputSelect'
import InputLabel from '../addressInputs/InputLabel'
import {
  getField,
  getListOfOptions,
  getDependentFields,
} from '../selectors/fields'
import reduce from 'lodash/reduce'

class SelectLevel extends Component {
  constructor(props) {
    super(props)

    this.state = {
      field: this.getLevelField(props),
    }
  }

  getLevelField({ level, rules }) {
    return getField(rules.postalCodeLevels[level], rules)
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ field: this.getLevelField(nextProps) })
  }

  handleChange = value => {
    const { field: { name } } = this.state
    const { rules } = this.props

    const dependentFields = getDependentFields(name, rules)

    const cleanAddress = reduce(
      this.props.address,
      (cleanAddress, value, prop) => {
        if (dependentFields.indexOf(prop) !== -1) {
          cleanAddress[prop] = { value: null }
        }
        return cleanAddress
      },
      {}
    )

    this.props.onChangeAddress({
      ...cleanAddress,
      [name]: {
        ...cleanAddress[name],
        value,
      },
    })
  };

  handleBlur = () => {
    const { field } = this.state
    const { address, onChangeAddress } = this.props

    onChangeAddress({
      [field.name]: {
        ...address[field.name],
        visited: true,
      },
    })
  };

  render() {
    const { field } = this.state
    const { rules, address } = this.props

    return (
      <InputLabel field={field}>
        <InputSelect
          field={field}
          address={address}
          options={getListOfOptions(field, address, rules)}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
        />
      </InputLabel>
    )
  }
}

SelectLevel.propTypes = {
  level: PropTypes.oneOf([0, 1]),
  address: PropTypes.shape(AddressShapeWithValidation),
  rules: PropTypes.object.isRequired,
  onChangeAddress: PropTypes.func.isRequired,
}

export default SelectLevel
