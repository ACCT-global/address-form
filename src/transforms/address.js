import reduce from 'lodash/reduce'
import filter from 'lodash/filter'
import map from 'lodash/map'
import difference from 'lodash/difference'
import find from 'lodash/find'
import findIndex from 'lodash/findIndex'
import { getField } from '../selectors/fields'
import { validateChangedFields } from '../validateAddress'
import msk from 'msk'

export function addValidation(address) {
  return reduce(
    address,
    (newAddress, value, propName) => {
      newAddress[propName] = { value }
      return newAddress
    },
    {}
  )
}

export function removeValidation(address) {
  return reduce(
    address,
    (newAddress, { value }, propName) => {
      newAddress[propName] = value
      return newAddress
    },
    {}
  )
}

export function addNewField(address, fieldName, value) {
  return reduce(
    address,
    (newAddress, prop, propName) => {
      newAddress[propName] = {
        ...prop,
        [fieldName]: value,
      }
      return newAddress
    },
    {}
  )
}

export function removeField(address, fieldName) {
  return reduce(
    address,
    (newAddress, prop, propName) => {
      newAddress[propName] = { ...prop }
      delete newAddress[propName][fieldName]
      return newAddress
    },
    {}
  )
}

export function addDisabledToProtectedFields(fields, rules) {
  return reduce(
    fields,
    (newFields, prop, propName) => {
      const hasValue = prop && prop.value
      const isProtectField =
        rules.postalCodeProtectedFields &&
        rules.postalCodeProtectedFields.indexOf(propName) !== -1

      newFields[propName] = prop

      if (isProtectField && hasValue) {
        newFields[propName] = {
          ...prop,
          disabled: true,
        }
      }

      return newFields
    },
    {}
  )
}

export function handleMultipleValues(fields) {
  return reduce(
    fields,
    (newFields, prop, propName) => {
      const hasMultipleValues = prop.value.indexOf(';') !== -1

      newFields[propName] = prop

      if (hasMultipleValues) {
        newFields[propName] = {
          ...prop,
          value: null,
          valueOptions: prop.value.split(';'),
        }
      }

      return newFields
    },
    {}
  )
}

export function maskFields(addressFields, rules) {
  return reduce(
    addressFields,
    (newAddressFields, prop, propName) => {
      const fieldRule = getField(propName, rules)

      newAddressFields[propName] = prop

      if (fieldRule && fieldRule.mask) {
        newAddressFields[propName] = {
          ...prop,
          ...(prop.value ? { value: msk(prop.value, fieldRule.mask) } : {}),
        }
      }

      return newAddressFields
    },
    {}
  )
}

export function addFocusToNextInvalidField(fields, rules) {
  const invalidFilledField = getFirstInvalidFilledField(fields, rules)

  if (invalidFilledField) {
    const { fieldName, field } = invalidFilledField

    return {
      ...fields,
      [fieldName]: field,
    }
  }

  const requiredField = getFirstRequiredFieldNotFilled(fields, rules)

  if (requiredField) {
    const { fieldName, field } = requiredField

    return {
      ...fields,
      [fieldName]: field,
    }
  }

  return fields
}

function getFirstInvalidFilledField(fields, rules) {
  const allFieldsVisited = addNewField(fields, 'visited', true)
  const validatedFields = validateChangedFields(
    allFieldsVisited,
    allFieldsVisited,
    rules
  )

  const orderedValidatedFields = orderFieldsByRules(validatedFields, rules)

  const firstInvalidField = find(
    orderedValidatedFields,
    field => field && field.valid === false
  )

  if (firstInvalidField) {
    return {
      fieldName: firstInvalidField.name,
      field: {
        ...validatedFields[firstInvalidField.name],
        focus: true,
      },
    }
  }

  return null
}

function orderFieldsByRules(fields, rules) {
  return reduce(
    fields,
    (acc, field, fieldName) => {
      const index = findIndex(
        rules.fields,
        ruleField => ruleField.name === fieldName
      )
      if (index === -1) return acc

      acc[index] = { ...field, name: fieldName }
      return acc
    },
    Array(Object.keys(fields).length)
  )
}

function getFirstRequiredFieldNotFilled(fields, rules) {
  const requiredFields = filter(rules.fields, field => field.required)
  const requiredFieldsNames = map(requiredFields, field => field.name)

  const fieldsNames = Object.keys(fields)
  const requiredFieldNotFilled = difference(requiredFieldsNames, fieldsNames)

  if (requiredFieldNotFilled && requiredFieldNotFilled.length > 0) {
    const nextRequiredFieldName = requiredFieldNotFilled[0]

    return { fieldName: nextRequiredFieldName, field: { focus: true } }
  }

  return null
}
