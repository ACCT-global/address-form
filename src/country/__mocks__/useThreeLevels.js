import { THREE_LEVELS } from '../../constants'
import { thirdLevelPostalCodes } from '../../transforms/postalCodes'
import {
  getOneLevel,
  getTwoLevels,
  getThreeLevels,
} from '../../transforms/addressFieldsOptions.js'

const countryData = {
  Beni: {
    Cercado: {
      Paititi: '10000',
      Trinidad: '10001',
      Yucumo: '10002',
    },
    Mamore: {
      'San Ramon': '10200',
    },
  },
  Cochabamba: {
    Arani: {
      Arani: '30000',
      Vacas: '30001',
    },
    Chapare: {
      Colomi: '30500',
      Epizana: '30501',
      Eterazama: '30502',
    },
  },
}

export default {
  postalCodeFrom: THREE_LEVELS,
  postalCodeLevels: ['state', 'city', 'neighborhood'],
  thirdLevelPostalCodes: thirdLevelPostalCodes(countryData),
  fields: [
    {
      name: 'street',
      label: 'street',
      required: true,
      size: 'xlarge',
    },
    {
      name: 'number',
      label: 'number',
      required: true,
      size: 'mini',
    },
    {
      name: 'complement',
      label: 'complement',
      size: 'large',
    },
    {
      name: 'state',
      label: 'department',
      required: true,
      size: 'large',
      level: 1,
      options: getOneLevel(countryData),
    },
    {
      name: 'city',
      label: 'province',
      required: true,
      size: 'large',
      level: 2,
      basedOn: 'state',
      optionsMap: getTwoLevels(countryData),
    },
    {
      name: 'neighborhood',
      label: 'city',
      required: true,
      level: 3,
      basedOn: 'city',
      optionsMap: getThreeLevels(countryData),
    },
  ],
}
