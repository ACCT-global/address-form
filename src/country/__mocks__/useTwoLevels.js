import { TWO_LEVELS } from '../../constants'
import { secondLevelPostalCodes } from '../../transforms/postalCodes'
import {
  getOneLevel,
  getTwoLevels,
} from '../../transforms/addressFieldsOptions'

const countryData = {
  'I Región': {
    Camiña: '1150000',
    Colchane: '1160000',
    Huara: '1140000',
    Iquique: '1100000',
    Pica: '1170000',
  },
  'II Región': {
    Antofagasta: '1240000',
    Calama: '1390000',
  },
}

export default {
  postalCodeFrom: TWO_LEVELS,
  postalCodeLevels: ['state', 'neighborhood'],
  secondLevelPostalCodes: secondLevelPostalCodes(countryData),
  fields: [
    {
      name: 'street',
      label: 'street',
      required: true,
      size: 'xlarge',
    },
    {
      name: 'number',
      label: 'number',
      required: true,
      size: 'mini',
    },
    {
      name: 'complement',
      label: 'complement',
      size: 'large',
    },
    {
      name: 'state',
      label: 'region',
      required: true,
      size: 'large',
      isUpperCase: false,
      options: getOneLevel(countryData),
    },
    {
      name: 'neighborhood',
      label: 'community',
      required: true,
      size: 'large',
      level: 2,
      basedOn: 'state',
      optionsMap: getTwoLevels(countryData),
    },
  ],
}
