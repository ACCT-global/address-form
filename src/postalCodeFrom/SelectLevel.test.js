import React from 'react'
import renderer from 'react-test-renderer'
import SelectLevel from './SelectLevel'
import { mount } from 'enzyme'
import useTwoLevels from '../country/__mocks__/useTwoLevels'
import useThreeLevels from '../country/__mocks__/useThreeLevels'
import address from '../__mocks__/newAddress'

describe('SelectLevel', () => {
  it('show state options', () => {
    const tree = renderer
      .create(
        <SelectLevel
          level={0}
          address={address}
          rules={useThreeLevels}
          onChangeAddress={jest.fn()}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('show city options', () => {
    const tree = renderer
      .create(
        <SelectLevel
          level={1}
          address={address}
          rules={useThreeLevels}
          onChangeAddress={jest.fn()}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should clear dependent fields when parent field changes', () => {
    const handleChange = jest.fn()
    const wrapper = mount(
      <SelectLevel
        level={0}
        address={{
          ...address,
          state: { value: 'II Región' },
          neighborhood: { value: 'Antofagasta' },
        }}
        rules={useTwoLevels}
        onChangeAddress={handleChange}
      />
    )

    const event = { target: { value: 'I Región' } }
    wrapper.find('select[name="state"]').simulate('change', event)

    expect(handleChange).toHaveBeenCalledWith({
      ...address,
      state: { value: 'I Región' },
      city: { value: null },
    })
  })

  it('should clear dependent fields when parent field changes', () => {
    const handleChange = jest.fn()
    const wrapper = mount(
      <SelectLevel
        level={0}
        address={{
          ...address,
          state: { value: 'Beni' },
          city: { value: 'Cercado' },
          neighborhood: { value: 'Yucumo' },
        }}
        rules={useThreeLevels}
        onChangeAddress={handleChange}
      />
    )

    const event = { target: { value: 'Cochabamba' } }
    wrapper.find('select[name="state"]').simulate('change', event)

    expect(handleChange).toHaveBeenCalledWith({
      ...address,
      state: { value: 'Cochabamba' },
      city: { value: null },
      neighborhood: { value: null },
    })
  })
})
