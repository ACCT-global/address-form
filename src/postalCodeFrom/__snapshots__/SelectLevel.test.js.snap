// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectLevel show city options 1`] = `
<p
  className="input required text"
>
  <label>
    province
  </label>
  <select
    className="input-large"
    name="city"
    onBlur={[Function]}
    onChange={[Function]}
    value=""
  >
    <option
      disabled={undefined}
      value=""
    />
  </select>
</p>
`;

exports[`SelectLevel show state options 1`] = `
<p
  className="input required text"
>
  <label>
    department
  </label>
  <select
    className="input-large"
    name="state"
    onBlur={[Function]}
    onChange={[Function]}
    value=""
  >
    <option
      disabled={undefined}
      value=""
    />
    <option
      value="Beni"
    >
      Beni
    </option>
    <option
      value="Cochabamba"
    >
      Cochabamba
    </option>
  </select>
</p>
`;
